returnn/datasets/concat_files.py

```diff
@@ -0,0 +1,613 @@
+        self._num_seqs = num_seqs
+
+    def get_num_seqs(self) -> int:
+        """num seqs for this sub epoch"""
+        self._lazy_wait_for_init_seq_order()
+        return self._num_seqs
+
+    def get_data_seq(self, seq_idx: int) -> Optional[DatasetSeq]:
+        """get data seq"""
+        self._lazy_wait_for_init_seq_order()
+        self.parent_conn.send(("get_data_seq", {"seq_idx": seq_idx}))
+        msg, data = self.parent_conn.recv()
+        assert msg == "data_seq"
+        return data
+
+    def exit(self, *, join: bool = True):
+        """exit"""
+        self._lazy_wait_for_init_seq_order()
+        self.parent_conn.send(("exit", {}))
+        if join:
+            self.worker_proc.join()
+        if self.exit_hook:
+            self.exit_hook()
+            self.exit_hook = None
+
+    def __del__(self):
+        # noinspection PyBroadException
+        try:
+            self.exit(join=False)
+        except Exception:
+            pass
+        else:
+            try_run(self.worker_proc.join)
+
+
+def _worker_proc_loop(
+    epoch: int,
+    buffer_size: int,
+    dataset_dict: Dict[str, Any],
+    parent_conn: mpConnection,
+):
+    if sys.platform == "linux":
+        with open("/proc/self/comm", "w") as f:
+            f.write(f"CFD worker {epoch}")
+
+    assert isinstance(epoch, int) and isinstance(buffer_size, int)
+    assert isinstance(dataset_dict, dict)
+    assert isinstance(parent_conn, mpConnection)
+
+    dataset = init_dataset(dataset_dict)
+
+    got_init_seq_order = False
+    cache: List[DatasetSeq] = []
+    next_seq_idx = 0
+
+    # noinspection PyShadowingNames
+    def _add_to_cache():
+        nonlocal next_seq_idx
+        if len(cache) >= buffer_size:
+            return False
+        if not dataset.is_less_than_num_seqs(next_seq_idx):
+            return False
+        dataset.load_seqs(next_seq_idx, next_seq_idx + 1)
+        seq_tag = dataset.get_tag(next_seq_idx)
+        features = {data_key: dataset.get_data(next_seq_idx, data_key) for data_key in dataset.get_data_keys()}
+        res = DatasetSeq(seq_idx=next_seq_idx, seq_tag=seq_tag, features=features)
+        cache.append(res)
+        next_seq_idx += 1
+        return True
+
+    # noinspection PyShadowingNames
+    def _get_from_cache(seq_idx: int) -> Optional[DatasetSeq]:
+        if not cache:
+            return None
+        if seq_idx > cache[-1].seq_idx:
+            return None
+        for seq in cache:
+            assert seq.seq_idx <= seq_idx
+            if seq.seq_idx == seq_idx:
+                return seq
+        assert False
+
+    # noinspection PyShadowingNames
+    def _get(seq_idx: int) -> Optional[DatasetSeq]:
+        if cache and seq_idx < cache[0].seq_idx:
+            raise Exception(
+                f"requested seq idx {seq_idx} is smaller than cache start {cache[0].seq_idx}, cannot go backwards"
+            )
+        res = _get_from_cache(seq_idx)
```

Comment:
If `_get` is called for every seq_idx from `0..SEQS_IN_FILE` we have O(n^2) search here.

Reply:
We clean up entries in the `cache` once we proceeded to higher seq idxs. (That's the protocol of the `Dataset` for `load_seqs`, that all earlier seqs are not used anymore, and `load_seqs` will always increase monotonically on seq idxs.)
See the code below:
```python
            elif msg == "get_data_seq":
                seq_idx = kwargs["seq_idx"]
                while cache and cache[0].seq_idx < seq_idx:
                    cache.pop(0)
                res = _get(seq_idx)
```

Reply:
> Seems like this would be a good fit for an implementation via a bounded queue (which the current impl kinda hand-rolls).

But it's not just a queue that I use here. I also need to be able to search through it and access random elements in it.

Or you mean to make the `pop(0)` more efficient? Yea, I was thinking about that, but I wanted to keep it simple, and specifically to use only Python stdlib for this here. Or what datastructure in Python stdlib do you think to use instead?

> Do you (personally) like async/await in python?

No, mostly because of the [function color argument](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/).

Reply:
> I also need to be able to search through it and access random elements in it.

I thought the access would be monotonic and that we might skip seqs but never require true random access?

> Or you mean to make the pop(0) more efficient? Yea, I was thinking about that, but I wanted to keep it simple, and specifically to use only Python stdlib for this here. Or what datastructure in Python stdlib do you think to use instead?

Yeah it's fine, no real worries here. That tiny bit of copying won't matter much I think.

> No, mostly because of the [function color argument](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/).

👍

I think it can be helpful for the concurrency it enables, and with the message passing pattern you implemented here I would say it would also fit very well (have one future poll the dataset, the other one manage communication w/ the parent process and multiplex it all on one reactor) but yeah function coloring is a thing. No need for changes here though, it's all simple and straightforward enough!